lib/gc/Transforms/LowerTileVectorPass.cpp

```diff
@@ -0,0 +1,680 @@
+/// it like other operation in the future.
+///
+/// TODO: Need to support upstream to handle empty vector size. Currently
+/// upstream folks don't allow me to do this. It's weird, I can't find reason.
+struct TensorUnpackConvertVectorPass : public RewritePattern {
+
+  explicit TensorUnpackConvertVectorPass(MLIRContext *context)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, context) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+
+    auto tensorUnPackOp = dyn_cast<tensor::UnPackOp>(op);
+    if (!tensorUnPackOp)
+      return rewriter.notifyMatchFailure(op, "Not expected operations.");
+
+    auto resultTy = cast<ShapedType>(op->getResultTypes()[0]);
+    // TODO: Need to support upstream to handle empty vector size. Currently
+    // upstream folks don't allow me to do this.
+    ArrayRef<int64_t> inputShape = resultTy.getShape();
+    SmallVector<bool, 5> targetVecDims(inputShape.size(), false);
+
+    if (failed(linalg::vectorize(rewriter, op,
+                                 /*inputVectorSizes=*/inputShape.vec(),
+                                 /*inputScalableVecDims=*/targetVecDims, false,
+                                 false))) {
+      return rewriter.notifyMatchFailure(op, "Fail to vectorize.");
+    }
+    return success();
+  }
+};
+
+/// Some tensor operation lowering to vector.
+///
+/// Currently support expand_shape, collapse_shape and concat_shape.
+/// May need support other operation in the future.
+struct TensorOpConvertVectorPass : public RewritePattern {
+private:
+  SmallVector<int64_t> inputVectorSizes;
+
+public:
+  explicit TensorOpConvertVectorPass(MLIRContext *context,
+                                     ArrayRef<int64_t> inputVectorSizes = {})
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, context),
+        inputVectorSizes(inputVectorSizes) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+
+    bool is_target = isRequiredTensorOp(op);
+    if (!is_target)
+      return rewriter.notifyMatchFailure(op, "Not expected operations.");
+
+    if (failed(convert2TargetOperation(rewriter, op, inputVectorSizes))) {
+      return rewriter.notifyMatchFailure(op, "Fail to vectorize.");
+    }
+    return success();
+  }
+};
+
+/// Patterns that lower to tile (virtual) vector.
+void populateLowerToTileVectorPatterns(RewritePatternSet &patterns) {
+  patterns.add<OperationConvertTileVectorPass<linalg::LinalgOp>,
+               OperationConvertTileVectorPass<tensor::PackOp>>(
+      patterns.getContext());
+  patterns.add<TensorUnpackConvertVectorPass>(patterns.getContext());
+  patterns.add<TensorOpConvertVectorPass>(patterns.getContext());
+}
+
+/// LowerToTileVectorPass is a pass that lowers operations to tile (virtual)
+/// vector. We must aware that this pass do not support dynamic shape currently.
+struct LowerTileVectorPass
```

Comment:
Better to name it as `LowerToTileVectorPass`, to be consistent with the class annotation in upper line.

Reply:
Reasonable~ Will unify them.